engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -477,15 +477,15 @@ private void validateConstraintsForCurrentGroup(ValidationContext<?> validationC
 	}
 
 	private <U> boolean validateConstraintsForSingleDefaultGroupElement(ValidationContext<?> validationContext, ValueContext<U, Object> valueContext, final Map<Class<?>, Class<?>> validatedInterfaces,
-			Class<? super U> clazz, Set<MetaConstraint<?>> metaConstraints, Group defaultSequenceMember) {
+			Class<? super U> clazz, Map<ConstraintLocation, Set<MetaConstraint<?>>> metaConstraintsByLocation, Group defaultSequenceMember) {
```

Comment:
I think you did the change the risky way (but i agree it's cute :)). And indeed a TCK test is broken.

I think you should still call this `validateConstraintsForSingleDefaultGroupElement()` method once per location. I think it's going to be more code changes to do it the way I think it should be but it will be safer.

IIRC I posted the pseudo code I think we should use in the JIRA issue. And it's in line with my other comment about making the `appendNode()` only once if we can.